tests/stack/test_api.py

```diff
@@ -529,3 +548,69 @@ def test_download_resource(self, stack_api):
+        stack_api._update_job = mock.MagicMock()
+        stack_api._update_job.return_value = 12345
+        stack_api._put_job = mock.MagicMock()
+        stack_api._put_job.return_value = 12345
+        stack_api.jobs_client.get_job = mock.MagicMock()
+        stack_api.jobs_client.get_job.return_value = {}
+
+        stack_api.workspace_client.import_workspace = mock.MagicMock()
+        stack_api.workspace_client.import_workspace_dir = mock.MagicMock()
+        stack_api.workspace_client.client.get_status = mock.MagicMock()
+        stack_api.workspace_client.client.get_status.return_value = {}
+
+        stack_api.dbfs_client.cp = mock.MagicMock()
+        stack_api.dbfs_client.client = mock.MagicMock()
+        stack_api.dbfs_client.client.get_status.return_value = {}
+
+        # Setup filesystem for workspace and dbfs resources
+        test_stack = copy.deepcopy(TEST_STACK)
+        for resource in test_stack[api.STACK_RESOURCES]:
+            resource_service = resource[api.RESOURCE_SERVICE]
+            resource_properties = resource[api.RESOURCE_PROPERTIES]
+            curr_source_path = resource_properties.get('source_path', '')
+            resource_properties.update(
+                {'source_path': os.path.join(tmpdir.strpath, curr_source_path)})
+            if resource_service == api.WORKSPACE_SERVICE:
+                if resource_properties['object_type'] == workspace_api.NOTEBOOK:
+                    os.makedirs(os.path.dirname(resource_properties['source_path']))
+                    with open(resource_properties['source_path'], 'w') as f:
+                        f.write("print('test')\n")
+                if resource_properties['object_type'] == workspace_api.DIRECTORY:
+                    os.makedirs(resource_properties['source_path'])
+            elif resource_service == api.DBFS_SERVICE:
+                if resource_properties['is_dir']:
+                    os.makedirs(resource_properties['source_path'])
+                else:
+                    with open(resource_properties['source_path'], 'w') as f:
+                        f.write("print('test')\n")
+
+        # Make sure inputted stack config and stack status is valid.
+        stack_api._validate_config(test_stack)
+        stack_api._validate_status(TEST_STATUS)
+
+        # Run deploy command on stack and validate stack status.
+        new_stack_status = stack_api.deploy_config(test_stack)
+        stack_api._validate_status(new_stack_status)
+        new_stack_status = stack_api.deploy_config(test_stack, stack_status=TEST_STATUS)
+        stack_api._validate_status(new_stack_status)
+        new_stack_status = stack_api.deploy_config(test_stack, stack_status=TEST_STATUS,
+                                                   overwrite=True)
+        stack_api._validate_status(new_stack_status)
+
+    def test_end_to_end_download(self, stack_api):
+        """
+            The stack API should not go through any validation exceptions for any resource types.
```

Comment:
Same here. It makes more sense to test "it should throw if the input is not expected" rather than testing "no exception if the input is expected". This test is also not end-to-end because there is no guarantee that the cli will perform the correct actions.
We should verify the correct REST API endpoints are called (with good parameters) after deploy/download are invoked, if we want to do an end to end tests.